### PR TITLE
android: sonatype maven release script

### PR DESCRIPTION
**Pull request by buildbreaker**
Maven public releases through sonatype requires the following artifacts:
1. `aar` or `jar` artifact
2. `pom.xml` which describes the artifact
3. `-sources.jar` which contains the code sources
4. `-javadoc.jar` which contains the java docs for the APIs
5. `*.asc` for each of the above signed (`gpg -ab "FILE"`)by a GPG key which sonatype knows about in order to verify the artifacts are uploaded by a trusted source

This means we have the following 8 files to upload:
Artifacts:
```
    dist/envoy.aar
    dist/envoy-pom.xml
    dist/envoy-sources.jar
    dist/envoy-javadoc.jar
```
GPG signed:
```
    dist/envoy.aar.asc
    dist/envoy-pom.xml.asc
    dist/envoy-sources.jar.asc
    dist/envoy-javadoc.jar.asc
```

The sonatype release process can be found [here](https://help.sonatype.com/repomanager2/staging-releases/staging-overview). In summary, we have:
1. Create a staging repository using a profile id (a profile id is associated with a namespace. Specially we care about `io.envoyproxy`)
2. Upload files to a specific staging repository (the one we created from step 1)
3. Close a specific staging repository
4. Release a specific staging repository (or alternatively drop a staging repository)

The release steps are the following:
1. First we sign all the files for uploading. The GPG signing was taken from [protoc-gen-validate](https://github.com/envoyproxy/protoc-gen-validate/blob/8ecc70e2e94e215f0d3f20c11325210871ee9204/.circleci/config.yml#L26-L28)
2. Use our python script to create a staging repository, upload, close, and release

References:
- [Create/upload/close a staging repository](https://support.sonatype.com/hc/en-us/articles/213465868-Uploading-to-a-Staging-Repository-via-REST-API)
- [List/release/drop a staging repository](https://support.sonatype.com/hc/en-us/articles/213465448-Automatically-dropping-old-staging-repositories)

Signed-off-by: Alan Chiu <achiu@lyft.com>

For an explanation of how to fill out the fields, please see the relevant section
in [PULL_REQUESTS.md](https://github.com/envoyproxy/envoy/blob/master/PULL_REQUESTS.md)

Description: android: sonatype maven release script 
Risk Level: low
Testing: local/ci
Docs Changes: n/a
Release Notes: n/a
[Optional Fixes #Issue]
[Optional Deprecated:]

### CODE REVIEW

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+          GPG_KEY: ${{ secrets.gpg_key }}
+          GPG_KEY_NAME: ${{ secrets.gpg_key_name }}
+          GPG_PASSPHRASE: ${{ secrets.gpg_passphrase }}
+          ENVOY_PROXY_PROFILE_ID: ${{ secrets.envoy_proxy_profile_id }}
+        run: |
+          # Sign and deploy to sonatype with the specified profile id
+          mvn -s bazel/settings.xml \
+            deploy:deploy-file \
+            -f dist/envoy-pom.xml \
+            -Dfile=envoy.aar \
```
Comment (buildbreaker) [inline .github/workflows/release.yml]: [ERROR] Failed to execute goal org.apache.maven.plugins:maven-deploy-plugin:3.0.0-M1:deploy-file (default-cli) on project envoy: /home/runner/work/envoy-mobile/envoy-mobile/dist/dist/envoy.aar not found. -> [Help 1]

File: .github/workflows/release.yml [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+          git config --global user.email "envoy-bot@users.noreply.github.com"
+          git config --global user.name "envoy-bot"
+      - name: 'Create sonatype staging profile id'
+        env:
+          SONATYPE_USER: ${{ secrets.sonatype_user }}
+          SONATYPE_PASSWORD: ${{ secrets.sonatype_password }}
+          ENVOY_PROXY_PROFILE_ID: ${{ secrets.envoy_proxy_profile_id }}
+        run: |
+          # Open release candidate with a specified profile id using `current_release_tag`
+          repository_id=$(curl -X POST -d '{"data":{"description":""}}' -u $SONATYPE_USER:$SONATYPE_PASSWORD -H "Content-Type:application/json" https://oss.sonatype.org/service/local/staging/profiles/$ENVOY_PROXY_PROFILE_ID/start | cut -d ":" -f3 | cut -d "}" -f1)
```
Comment (buildbreaker) [inline .github/workflows/release.yml]: fix this

File: .github/workflows/release.yml [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+          # Import gpg keys and warm the passphrase to avoid the gpg
+          # passphrase prompt when initating a deploy
+          # `--pinentry-mode=loopback` could be needed to ensure we
+          # suppress the gpg prompt
+          echo $GPG_KEY | base64 --decode > signing-key
+          gpg --passphrase $GPG_PASSPHRASE --batch --import signing-key
+          shred signing-key
+
+          gpg --pinentry-mode=loopback --passphrase $GPG_PASSPHRASE -ab dist/envoy.aar
+          gpg --pinentry-mode=loopback --passphrase $GPG_PASSPHRASE -ab dist/envoy-pom.xml
```
[the remark this answers is not in the document]
Comment (buildbreaker) [reply .github/workflows/release.yml]: Doesn't seem like `--batch` worked for me: `gpg --passphrase testpass --batch -ab WORKSPACE VERSION`

File: .github/workflows/release.yml [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+        env:
+          SONATYPE_USER: ${{ secrets.sonatype_user }}
+          SONATYPE_PASSWORD: ${{ secrets.sonatype_password }}
+          ENVOY_PROXY_PROFILE_ID: ${{ secrets.envoy_proxy_profile_id }}
+        run: |
+          # Open release candidate with a specified profile id using `current_release_tag`
+          export READWRITE_USER=$SONATYPE_USER
+          export READWRITE_API_KEY=$SONATYPE_PASSWORD
+
+          current_release_tag=$(git describe --tags `git rev-list --tags --max-count=1`)
```
[the remark this answers is not in the document]
Comment (buildbreaker) [reply .github/workflows/release.yml]: I can look into if/how github actions passes down the release tag but there is a way for us to only execute an action on a release (I commented out the trigger up top)

File: dist/upload_maven.py [hunk 5]
```diff
@@ -15,123 +13,250 @@
...
+        return urlopen(request)
+    except HTTPError as e:
+        if retries > 0 and e.code >= 500:
+            print("Retrying request for {}".format(request.url))
+            return urlopen_retried(request, retries - 1)
+        else:
+            raise e
+
+
+def _create_staging_repository(profile_id):
```
[the remark this answers is not in the document]
Comment (buildbreaker) [reply dist/upload_maven.py]: How about I rename the file to be `sonatype_nexus_upload` instead?

File: dist/upload_maven.py [hunk 6]
```diff
@@ -15,123 +13,250 @@
...
 
-    os.makedirs(path)
+    try:
+        request = Request(url)
+
+        request.add_header("Authorization", "Basic {}".format(ENCODED_CREDENTIALS))
+        request.add_header("Content-Type", "application/json")
+        request.add_data(json.dumps(data))
+        request.get_method = lambda: "PUT"
+        urlopen_retried(request)
```
[the remark this answers is not in the document]
Comment (buildbreaker) [reply dist/upload_maven.py]: For all the request retries?

File: dist/upload_maven.py [hunk 7]
```diff
@@ -15,123 +13,250 @@
...
+        request.get_method = lambda: "POST"
+        urlopen_retried(request)
+    except Exception as e:
+        raise e
 
 
 def _build_parser():
     parser = argparse.ArgumentParser()
-    parser.add_argument("--version", default="LOCAL-SNAPSHOT")
+    parser.add_argument("--profile_id", required=True,
```
[the remark this answers is not in the document]
Comment (buildbreaker) [reply dist/upload_maven.py]: Let's rename the file so it's implied that it's sonatype specific

File: dist/upload_maven.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (buildbreaker) [reply dist/upload_maven.py]: Adding a 1s delay

File: .github/workflows/release.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (buildbreaker) [reply .github/workflows/release.yml]: I can get a `GITHUB_REF` but that includes branches?

File: .github/workflows/release.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (buildbreaker) [reply .github/workflows/release.yml]: We can also lift this into a script in the future :)

File: .github/workflows/release.yml [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+          export PATH=/usr/lib/llvm-8/bin:$PATH
+          export CC=clang
+          export CXX=clang++
+          export ANDROID_NDK_HOME=/usr/local/lib/android/sdk/ndk-bundle
+
+          # Build and create the artifacts
+          current_release_tag=$(git describe --tags `git rev-list --tags --max-count=1`)
+          ./bazelw build \
+            --config=release-android \
+            --fat_apk_cpu=x86 \
```
Comment (buildbreaker) [inline .github/workflows/release.yml]: need to change to all architectures

File: .github/workflows/release.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (artem-zinnatullin) [reply .github/workflows/release.yml]: 👍

File: .github/workflows/release.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (artem-zinnatullin) [reply .github/workflows/release.yml]: You should be able to do with git commands, something like this: `git describe --exact-match HEAD`

File: BUILD [hunk 9]
```diff
@@ -70,6 +70,10 @@ chmod 755 dist/envoy.aar
 chmod 755 dist/envoy-pom.xml
 chmod 755 dist/envoy-javadoc.jar
 chmod 755 dist/envoy-sources.jar
+sha256sum dist/envoy.aar | cut -d " " -f1 > dist/envoy.aar.sha256
```
Comment (artem-zinnatullin) [inline BUILD]: Maybe do it in Python script instead of doing `cut -d " "` stuff

File: dist/sonatype_nexus_upload.py [hunk 10]
```diff
@@ -0,0 +1,321 @@
...
+        return urlopen(request)
+    except HTTPError as e:
+        if max_retries > 0 and e.code >= 500:
+            print("[{retry_attempt}/{max_retries} Retry attempt] Retrying request. Received error code {code}".format(
+                retry_attempt=attempt,
+                max_retries=max_retries,
+                code=e.code
+            ),
+                file=sys.stderr)
+            time.sleep(delay_sec)
```
Comment (artem-zinnatullin) [inline dist/sonatype_nexus_upload.py]: I'd print warning about sleep too :)

File: .github/workflows/release.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (buildbreaker) [reply .github/workflows/release.yml]:
```
git describe --tags --abbrev=0
```

File: BUILD (hunk 9 above)
Comment (buildbreaker) [reply BUILD]: I can actually just have the python script generate the hashes

Comment (artem-zinnatullin) [reply BUILD]: That's what I meant :)

File: dist/sonatype_nexus_upload.py [hunk 11]
```diff
@@ -0,0 +1,347 @@
...
+
+        # Upload files using the staging_id, close the staging repository, and release
+        # If an error occurs, we will attempt to drop the repository. The script will
+        # need to be re-run to initiate another upload attempt
+        try:
+            print("Uploading files...")
+            sha256_files = _create_sha256_files(args.files)
+            uploaded_file_count = _upload_files(staging_id, version, args.files, args.signed_files, sha256_files)
+            if uploaded_file_count > 0:
+                print("Uploading files complete!")
```
Comment (rebello95) [inline dist/sonatype_nexus_upload.py]: These are great for documentation 👌

**Pull request merged**